### PR TITLE
added new batching rules for `_scaled_dot_product_efficient_attention`

### LINKED ISSUE #102457: Batching rule for `aten::_scaled_dot_product_efficient_attention`
### 🚀 The feature, motivation and pitch

Hi, I am trying to take batched gradients of a vector output given by `_scaled_dot_product_efficient_attention` but saw the error 

```
/site-packages/optimum/bettertransformer/models/attention.py:56: UserWarning: There is a performance drop because we have not yet implemented the batching rul
e for aten::_scaled_dot_product_efficient_attention. Please file us an issue on GitHub so that we can prioritize its implementation. (Triggered internally at ../aten/src/ATen/functorch/BatchedFallback.cpp:82.)
  sdpa_result = torch.nn.functional.scaled_dot_product_attention
```

when running my code. Implementing this would really increase throughput in our application! I would also be happy to take a stab at implementing it, if there is a document describing what I need to do at a high level.

### Alternatives

_No response_

### Additional context

_No response_

cc @ezyang @gchanan @zou3519 @kadeng @jbschlosser @bhosmer @cpuhrsch @erichan1 @drisspg @mikaylagawarecki @Chillee @samdow @kshitij12345 @janeyx99

### PR DESCRIPTION
**Pull request by lengstrom**
Fixes #102457 

(previously submitted PR at https://github.com/pytorch/pytorch/pull/103159; had some git issues so submitted a new one; this PR no longer deletes the chunked attention output code)

### CODE REVIEW

File: aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp [hunk 1]
```diff
@@ -280,6 +280,62 @@ static void expect_at_least_rank(
...
+  const auto key_ = key_bdim.has_value() ? reshape_dim_into(*key_bdim, 0, key) : key;
+  const auto value_ = value_bdim ? reshape_dim_into(*value_bdim, 0, value) : value;
+
+  const auto res = at::_scaled_dot_product_efficient_attention(query_, key_,
+      value_, compute_log_sumexp, dropout_p, is_causal, scale);
+  const auto batch_size = query.size(*query_bdim);
+  const auto res0_ = reshape_dim_outof(0, batch_size, std::get<0>(res));
+  const auto res1_ = reshape_dim_outof(0, batch_size, std::get<1>(res));
+  const auto res2_ = reshape_dim_outof(0, batch_size, std::get<2>(res));
+  const auto res3_ = reshape_dim_outof(0, batch_size, std::get<2>(res));
```
Comment (Skylion007) [inline aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp:302]: Looks like a typo.  This is the same as res2.

File: aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp [hunk 2]
```diff
@@ -280,6 +280,62 @@ static void expect_at_least_rank(
...
+  bool compute_log_sumexp,
+  double dropout_p,
+  bool is_causal,
+  c10::optional<double> scale
+) {
+  const auto query_ = query_bdim.has_value() ? reshape_dim_into(*query_bdim, 0, query) : query;
+  const auto key_ = key_bdim.has_value() ? reshape_dim_into(*key_bdim, 0, key) : key;
+  const auto value_ = value_bdim ? reshape_dim_into(*value_bdim, 0, value) : value;
+
+  const auto res = at::_scaled_dot_product_efficient_attention(query_, key_,
```
Comment (Skylion007) [inline aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp:296]: You may be able to make this a bit more readable by using structured bindings in C++17 to extract the values directly from the std::tuple.

**Review by Skylion007 — COMMENTED** [review]
Left some comments

File: aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp (hunk 1 above)
Comment (lengstrom) [reply aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp:302]: oops thanks!

Comment (lengstrom) [reply aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp]: shoudl be fixed in new commit

**Comment by lengstrom** [comment]
hi! are there other modifications I can make to this PR to merge it in?

File: aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp [hunk 3]
```diff
@@ -280,6 +280,62 @@ static void expect_at_least_rank(
       rank, " dimensions instead.");
 }
 
+fourOutputs _scaled_dot_product_efficient_attention_batch_rule(
```
Comment (zou3519) [inline aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp:283]: In order to test this, we should add a new OpInfo that has dropout_p = 0. This will give us automated tests for this operator (nn.functional.scaled_dot_product_attention); you'll be able to run `pytest test/functorch/ -v -k "scaled_dot_product_attention".


There is an existing OpInfo that has dropout_p > 0; our test handler can't handle randomness. You should be able to just copy-paste https://github.com/pytorch/pytorch/blob/b2277075b0fe5cf085d369a313863f64c6fb50c3/torch/testing/_internal/common_methods_invocations.py#L13240 and set dropout_p = 0.

**Review by zou3519 — COMMENTED** [review]
The code looks fine to me, but we should add testing for it. The automated testing we have is pretty good and will exercise all the cases.

**Comment by lengstrom** [comment]
awesome thanks! i added the test cases as you said and they fail; i will try to fix soon

**Comment by malbergo** [comment]
Did this ever get merged? Currently one cannot use vmap with _scaled_dor_product_efficient_attention.

**Comment by JamesTheZ** [comment]
This is still not merged till torch 2.3.1. Is there any plan to merge this function @zou3519
